server/index.ts

```diff
@@ -55,22 +55,61 @@ import {
 } from "../build/blog.js";
 import { findCurriculumPageBySlug } from "../build/curriculum.js";
 
-async function buildDocumentFromURL(url: string) {
-  const document = Document.findByURL(url);
-  if (!document) {
-    return null;
+async function buildDocumentFromURL(res, url: string) {
+  try {
+    console.time(`buildDocumentFromURL(${url})`);
+    const document = Document.findByURL(url);
+    if (!document) {
+      return null;
+    }
+    const documentOptions = {};
+    if (CONTENT_TRANSLATED_ROOT) {
+      // When you're running the dev server and build documents
+      // every time a URL is requested, you won't have had the chance to do
+      // the phase that happens when you do a regular `yarn build`.
+      document.translations = findTranslations(
+        document.metadata.slug,
+        document.metadata.locale
+      );
+    }
+    const built = await buildDocument(document, documentOptions);
+
+    if (built) {
+      return { doc: built?.doc, url };
+    } else if (
+      url.split("/")[1] &&
+      url.split("/")[1].toLowerCase() !== DEFAULT_LOCALE.toLowerCase() &&
+      !CONTENT_TRANSLATED_ROOT
+    ) {
+      // Such a common mistake. You try to view a URL that is not en-US but
+      // you forgot to set CONTENT_TRANSLATED_ROOT.
+      console.warn(
+        `URL is for locale '${
+          url.split("/")[1]
+        }' but CONTENT_TRANSLATED_ROOT is not set. URL will 404.`
+      );
+    }
+  } catch (error) {
+    console.error(`Error in buildDocumentFromURL(${url})`, error);
+    return res.status(500).json(JSON.stringify(error.toString()));
```

Comment:
With the current logic of `const doc = buildDocumentFromURL` we get into a weird situation where `doc` becomes a response object